doc/GUIDE.md

```diff
@@ -0,0 +1,910 @@
+whether the typedef is for a primitive type or a struct/union type respectively.
+
+Take for example the following `typedef` declarations:
+
+```c
+// mylib.h
+
+typedef int MyInt;
+
+struct Point {
+    int x;
+    int y;
+};
+typedef struct Point MyPoint;
+```
+
+`MyInt` is a `typedef` of the primitive type `int`, so it is translated by jextract as a
+`static final` layout field in the main header class:
+
+```java
+// mylib_h.java
+
+public static final OfInt MyInt = mylib_h.C_INT;
+```
+
+The `MyPoint` `typedef`, on the other hand, is a typedef for a struct, so it is translated
+as a separate class which extends the class that is generated for the `Point` struct:
+
+```java
+// MyPoint.java
+
+public class MyPoint extends Point { }
+```
+
+Through static inheritance, all the methods in the `Point` class are available through the
+`MyPoint` class as well.
+
+### Nested Types
+
+C allows variable declarations to have an inline anonymous type. For instance, if we
+have a struct such as this:
+
+```c
+// mylib.h
+
+struct Foo {
+    struct {
+        int baz;
+    } bar; // field of Foo
+
+    void (*cb)(void);
+};
+```
+
+Jextract generates a _nested_ struct and function pointer class for the `bar` and `cb`
+fields _inside of_ the class it generates for the `Foo` struct itself:
+
+```java
+// mylib_h.java
+
+public class Foo {
+    ...
+    public static class bar { ... }
+    public static final GroupLayout bar$layout() { ... }
+    public static final long bar$offset() { ... }
+
+    public static MemorySegment bar(MemorySegment struct) { ... }
+    public static void bar(MemorySegment struct, MemorySegment fieldValue) { ... }
+    ...
+    public static class cb { ... }
+    public static final AddressLayout cb$layout() { ... }
+    public static final long cb$offset() { ... }
+
+    public static MemorySegment cb(MemorySegment struct) { ... }
+    public static void cb(MemorySegment struct, MemorySegment fieldValue) { ... }
+    ...
+}
+```
+
+In both cases, the name of the nested class is the name of the field of the struct.
+
+Both fields also have the usual getter and setter, but note that the getter for the struct
+field returns a _reference_ to the memory inside the `Foo` struct that corresponds to the
+`bar` field. This means that writes to the returned memory segment will be visible in the
+enclosing struct instance as well:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment foo = Foo.allocate(arena);
+    MemorySegment bar = Foo.bar(foo); // reference to bar inside foo
+    System.out.println(Foo.bar.baz(bar));  // prints: 0
+
+    MemorySegment bar2 = Foo.bar.allocate(arena);
+    Foo.bar.baz(bar2, 42);
+    Foo.bar(foo, bar2);  // copies bar2 into foo
+    System.out.println(Foo.bar.baz(bar));  // prints: 42
+}
+```
+
+In the above snippet, note that the load of the `baz` field value on the last line will
+_see_ the update to the `bar` field of the `foo` instance on the line before.
+
```

Comment:
Perhaps we could also add a section (maybe in a followup PR) on fields/globals with array types, as jextract generates extra metadata (e.g. dimensions) and the Java type used for these accesses is MemorySegment (with bulk copy support on `set`)

Reply:
Yes, good point. This was missed

Reply:
let's save this for a followup, and get this stake in the ground first :) (it will probably also be easier to review a small section)